qi-lib/flow/core/compiler/deforest/cps.rkt

```diff
@@ -0,0 +1,371 @@
+       (cond
+         [(attribute blanket?)
+          (make-blanket-curry pre-arg
+                              post-arg
+                              max-args
+                              #'form-stx
+                              )]
+         [(attribute fine?)
+          (make-fine-curry arg min-args max-args #'form-stx)]
+         [else
+          (lambda (ctx name) #'(lambda (v) v))])]))
+
+  (define-syntax-class fsp
+    #:attributes (curry name contract prepare next)
+    (pattern range:fsp-range
+             #:attr name #''range
+             #:attr contract #'(->* (real?) (real? real?) any)
+             #:attr prepare #'range->cstream-prepare
+             #:attr next #'range->cstream-next
+             #:attr curry (make-producer-curry 1 3
+                                               range.blanket? #'range.pre-arg #'range.post-arg
+                                               range.fine? #'range.arg
+                                               range))
+    (pattern default:fsp-default
+             #:attr name #''list->cstream
+             #:attr contract #'(-> list? any)
+             #:attr prepare #'list->cstream-prepare
+             #:attr next #'list->cstream-next
+             #:attr curry (lambda (ctx name) #'(lambda (v) v)))
+    )
+
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  ;; Transformers
+
+  (define-syntax-class fst
+    #:attributes (next f state)
+    (pattern filter:fst-filter
+             #:attr f #'(filter.f)
+             #:attr next #'filter-cstream-next
+             #:attr state #'())
+    (pattern map:fst-map
+             #:attr f #'(map.f)
+             #:attr next #'map-cstream-next
+             #:attr state #'())
+    (pattern filter-map:fst-filter-map
+             #:attr f #'(filter-map.f)
+             #:attr next #'filter-map-cstream-next
+             #:attr state #'())
+    (pattern take:fst-take
+             #:attr f #'()
+             #:attr next #'take-cstream-next
+             #:attr state #'(take.n))
+    )
+
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  ;; Consumers
+
+  (define-syntax-class fsc
+    #:attributes (end)
+    (pattern foldr:fsc-foldr
+             #:attr end #'(foldr-cstream-next foldr.op foldr.init))
+    (pattern foldl:fsc-foldl
+             #:attr end #'(foldl-cstream-next foldl.op foldl.init))
+    (pattern list-ref:fsc-list-ref
+             #:attr end #'(list-ref-cstream-next list-ref.pos 'list-ref.name))
+    (pattern length:fsc-length
+             #:attr end #'(length-cstream-next))
+    (pattern empty?:fsc-empty?
+             #:attr end #'(empty?-cstream-next))
+    (pattern default:fsc-default
+             #:attr end #'(cstream-next->list))
+    )
+
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  ;; The pass
+
+  ;; Performs deforestation rewrite on the whole syntax tree.
+  (define-and-register-deforest-pass (deforest-pass ops ctx)
+    (syntax-parse (reverse ops)
```

Comment:
Would prefer using `define-and-register-pass` directly here rather than use a macro we don't anticipate another need for. To keep it modular, this `(syntax-parse ...`) bit could be moved into a locally-defined function.

Reply:
See above - this one will be also used in the named-let implementation in the (near) future.